### PR TITLE
PR: Fix extraneous reloading documents on project switching

### LINKED ISSUE #14787: Editor extraneously reloads files when switching projects
## Problem Description

When switching projects, the Editor seems to close and reopen the documents from the _first_ project, then switches to the new project and loads the documents for the new projects.

### What steps reproduce the problem?

1. Open a project with several files
2. Switch to another project
3. Observe the files from first project being closed and reopened, then the new project is loaded with its files.

### What is the expected output? What do you see instead?
When switching projects, the first project files should be closed and _not_ reopened.

## Versions

* Spyder version: 4.3.0.dev0 1c4c84e72
* Python version: 3.9.1 64-bit
* Qt version: 5.9.7
* PyQt5 version: 5.9.2
* Operating System: Darwin 19.6.0

### Dependencies

<details><summary>Dependencies</summary>

```
# Mandatory:
applaunchservices >=0.1.7       :  0.2.1 (OK)
atomicwrites >=1.2.0            :  1.4.0 (OK)
chardet >=2.0.0                 :  4.0.0 (OK)
cloudpickle >=0.5.0             :  1.6.0 (OK)
diff_match_patch >=20181111     :  20200713 (OK)
intervaltree >=3.0.2            :  3.1.0 (OK)
IPython >=7.6.0                 :  7.20.0 (OK)
jedi =0.17.2                    :  0.17.2 (OK)
jsonschema >=3.2.0              :  3.2.0 (OK)
keyring >=17.0.0                :  22.0.1 (OK)
nbconvert >=4.0                 :  6.0.7 (OK)
numpydoc >=0.6.0                :  1.1.0 (OK)
parso =0.7.0                    :  0.7.0 (OK)
pexpect >=4.4.0                 :  4.8.0 (OK)
pickleshare >=0.4               :  0.7.5 (OK)
psutil >=5.3                    :  5.8.0 (OK)
pygments >=2.0                  :  2.7.4 (OK)
pylint >=1.0                    :  2.6.0 (OK)
pyls >=0.36.2;<1.0.0            :  v4.2.1+130.g1c4c84e72 (OK)
pyls_black >=0.4.6              :  0.4.6 (OK)
pyls_spyder >=0.3.0             :  0.3.0 (OK)
qdarkstyle >=2.8                :  2.8.1 (OK)
qtawesome >=0.5.7               :  1.0.1 (OK)
qtconsole >=5.0.1               :  5.0.2 (OK)
qtpy >=1.5.0                    :  1.9.0 (OK)
rtree >=0.8.3                   :  0.9.4 (OK)
setuptools >=39.0.0             :  52.0.0.post20210125 (OK)
sphinx >=0.6.6                  :  3.4.3 (OK)
spyder_kernels >=1.10.1;<1.11.0 :  1.11.0.dev0 (OK)
textdistance >=4.2.0            :  4.2.0 (OK)
three_merge >=0.1.1             :  0.1.1 (OK)
watchdog >=0.10.3               :  1.0.2 (OK)
zmq >=17                        :  20.0.0 (OK)

# Optional:
cython >=0.21                   :  0.29.21 (OK)
matplotlib >=2.0.0              :  3.3.4 (OK)
numpy >=1.7                     :  1.19.2 (OK)
pandas >=1.1.1                  :  1.2.1 (OK)
scipy >=0.17.0                  :  1.6.0 (OK)
sympy >=0.7.3                   :  1.7.1 (OK)
```
</details>

change projects

### PR DESCRIPTION
**Pull request by mrclary**
## Description of Changes

* When switching projects, project close emit will call `editor.setup_open_files` only when not switching projects.
* Removed redundant code in `create_new_project`. This lead to multiple unnecessary calls to `sig_project_closed.emit` and `restart_consoles`.
* Removed unnecessary test for project close signal on create new project. This test fails because a project `root_path` is not set in the mocked dialogue window so a project is not _actually_ created and the active project is not closed. This test is unnecessary since `create_new_project` should not send `sig_project_closed.emit`, which is sent in `open_project` (if switching projects) or the `close_project` and is already tested.

### Issue(s) Resolved

Fixes #14787

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:
@mrclary

### CODE REVIEW

**Comment by mrclary** [comment]
@ccordoba12, this is ready for review.

**Comment by ccordoba12** [comment]
> When switching projects, project close emit will call editor.setup_open_files only when not switching projects.

So what happens when you switch projects? I mean, how the files of the project that is opened (after the current one is closed) are set in the editor?

**Comment by mrclary** [comment]
> > When switching projects, project close emit will call editor.setup_open_files only when not switching projects.
> 
> So what happens when you switch projects? I mean, how the files of the project that is opened (after the current one is closed) are set in the editor?

The editor.setup_open_files gets called with the sig_project_loaded.emit in the open_project method of the Projects plugin class. So basically, it gets called when a project gets closed, also when a projects gets opened; but when one switches a project, it was called twice (once with the sig_project_closed and once with the sig_project_loaded), causing the Editor to reload the files in the first project before loading the files in the second project. So, I made a switch in the sig_project_closed closed to distinguish between events where the project is being closed only, and events where the project is closed for switching to another project. In this way, the setup_open_files is always called, but only once.

File: spyder/plugins/projects/plugin.py [hunk 1]
```diff
@@ -213,18 +215,19 @@ def register(self):
                 lambda v: outline_explorer.update_all_editors())
         if self.main:
             self.sig_project_closed.connect(
-                lambda v: self.main.set_window_title())
+                lambda v, s: self.main.set_window_title())
```
Comment (ccordoba12) [inline spyder/plugins/projects/plugin.py:218]: Please revert this change. The thing is when not setting the signature in a signal, then it uses the first one as default (i.e. the same as before).

File: spyder/plugins/projects/plugin.py [hunk 2]
```diff
@@ -213,18 +215,19 @@ def register(self):
...
             self.sig_project_closed.connect(
-                lambda v: self.main.set_window_title())
+                lambda v, s: self.main.set_window_title())
         if self.completions:
             self.sig_project_closed.connect(
                 functools.partial(self.completions.project_path_update,
                                   update_kind=WorkspaceUpdateKind.DELETION,
                                   instance=self))
         if self.editor:
             self.sig_project_closed.connect(
```
Comment (ccordoba12) [inline spyder/plugins/projects/plugin.py:225]:
```suggestion
            self.sig_project_closed[bool].connect(
```

Here we're using the signal but with its second signature, so we can connect it to `setup_open_files`.

File: spyder/plugins/projects/plugin.py [hunk 3]
```diff
@@ -213,18 +215,19 @@ def register(self):
...
-                lambda v: self.editor.setup_open_files())
+                lambda v, s: (None if s
+                              else self.editor.setup_open_files()))
```
Comment (ccordoba12) [inline spyder/plugins/projects/plugin.py:227]: Please revert. Now only `self.sig_project_closed[bool]` is connected to this slot.

File: spyder/plugins/projects/plugin.py [hunk 4]
```diff
@@ -213,18 +215,19 @@ def register(self):
...
         if outline_explorer:
             self.sig_project_closed.connect(
-                lambda v: outline_explorer.update_all_editors())
+                lambda v, s: outline_explorer.update_all_editors())
```
Comment (ccordoba12) [inline spyder/plugins/projects/plugin.py:230]: Please revert.

File: spyder/plugins/projects/plugin.py [hunk 5]
```diff
@@ -234,7 +237,7 @@ def register(self):
             self.sig_project_loaded.connect(
                 lambda v: self.editor.set_current_project_path(v))
             self.sig_project_closed.connect(
-                lambda v: self.editor.set_current_project_path())
+                lambda v, s: self.editor.set_current_project_path())
```
Comment (ccordoba12) [inline spyder/plugins/projects/plugin.py:240]: Please revert.

File: spyder/plugins/projects/plugin.py [hunk 6]
```diff
@@ -467,7 +457,7 @@ def open_project(self, path=None, project=None, restart_consoles=True,
             # TODO: Don't emit sig_project_closed when we support
             # multiple workspaces.
             self.sig_project_closed.emit(
-                self.current_active_project.root_path)
+                self.current_active_project.root_path, True)
```
Comment (ccordoba12) [inline spyder/plugins/projects/plugin.py:460]: Please revert.

File: spyder/plugins/projects/plugin.py [hunk 7]
```diff
@@ -512,7 +502,7 @@ def close_project(self):
             self.set_conf('current_project_path', None)
             self.setup_menu_actions()
 
-            self.sig_project_closed.emit(path)
+            self.sig_project_closed.emit(path, False)
```
Comment (ccordoba12) [inline spyder/plugins/projects/plugin.py:505]:
```suggestion
            self.sig_project_closed.emit(path)
            self.sig_project_closed[bool].emit(True)
```

* The first signal needs to be emitted for things to work as before.
* As you can probably guess for the second signal, the emitted value is irrelevant. The important thing is its signature.

File: spyder/plugins/workingdirectory/plugin.py [hunk 8]
```diff
@@ -109,8 +109,8 @@ def register(self):
                 )
             )
 
-            projects.sig_project_closed[object].connect(
-                lambda path: self.chdir(
+            projects.sig_project_closed.connect(
+                lambda path, s: self.chdir(
```
Comment (ccordoba12) [inline spyder/plugins/workingdirectory/plugin.py:113]: Please revert.

**Review by ccordoba12 — COMMENTED** [review]
Thanks @mrclary for your help with this! I think by using `sig_project_closed` with an additional signature, your PR can be made much simpler, so all my suggestions below are oriented towards that.

**Comment by mrclary** [comment]
> Thanks @mrclary for your help with this! I think by using `sig_project_closed` with an additional signature, your PR can be made much simpler, so all my suggestions below are oriented towards that.

@ccordoba12, thanks for the suggestions! I learned a lot from that.

**Comment by ccordoba12** [comment]
> @ccordoba12, thanks for the suggestions! I learned a lot from that.

I'm really glad about it!

**Review by ccordoba12 — APPROVED** [review]
Looks good to me now, thanks @mrclary!
